### ISSUE TITLE
Search context: Add search context filter type for `searchContext` GQL query

### ISSUE THREAD

**Issue by vovakulikov** [sourcegraph/sourcegraph#34402]
At the moment we may have different types of search contexts. 

- Standard contexts
- Auto-created contexts
- query-based contexts

Unfortunately, we don't have filter functionality over the context in order to query only one type of context. For example, I want to query only query-based context (this is actually a real use case for part of Code Insights filter UI) with existing API I have to query all fields on the backend and make FE filtering. This implies a few perf problems - how can we handle cases when we have a lot of contexts? 

I suggest introducing a type filter filed over search context that we could query contexts by type 

```gql
query SearchBasedContexts {
   searchContexts(type: SearchContext.QueryBased) {
     id
     name
     ...
   }
}
```

**Comment by sourcegraph-bot-2** [comment]
Heads up @benvenker @lguychard - the "team/search-product" label was applied to this issue.

**Comment by sourcegraph-bot-2** [comment]
Heads up @joelkw @felixfbecker @vovakulikov @unclejustin - the "team/code-insights" label was applied to this issue.

**Comment by benvenker** [comment]
@novoselrok FYI this would really help out Code Insights and I think could enable some interesting things for us, like filtering/grouping contexts in the UI. Any thoughts on complexity here?

**Comment by felixfbecker** [comment]
@unclejustin @vovakulikov  Do I understand correctly that this would be implemented by the search team, not by us? (it's labelled both and in our iteration board)

**Comment by vovakulikov** [comment]
@felixfbecker I don't know actually, ping @novoselrok. 
For now, we implemented our search context filter with the existing state of our API but this may have perf problems for big customers.

**Comment by Joelkw** [comment]
@benvenker (cc @novoselrok ) I know you're at an offsite so no urgency, but sometime this week could you let us know your thoughts on this?

**Comment by novoselrok** [comment]
Yes, this makes sense. We can add a `type` filter to the `searchContexts` query that would support filtering by query-based contexts and "static" contexts. Auto-defined contexts are gathered through a separate query and I wouldn't add them as a separate type.

**Comment by vovakulikov** [comment]
@novoselrok 
> Auto-defined contexts are gathered through a separate query, and I wouldn't add them as a separate type

Just out of curiosity, why do we query them in a separate query? I thought it was done in this way because, in the `searchContexts` query, we don't have the ability to filter them and gather only auto-defined contexts. But if we would have a filter value for auto-defined contexts we didn't need to have a separate query, right?

**Comment by novoselrok** [comment]
> Just out of curiosity, why do we query them in a separate query?

It's because auto-defined contexts are not stored in the database and we generate them on the fly. That is why we can't mix them because they come from different "sources." 

I realize that having two queries is tricky from the client's perspective, but having them merged would be even more challenging in the backend.
